### ISSUE TITLE
Babel breaks async/await code using async npm package and promisify.

### ISSUE THREAD

**Issue by ealmansi** [babel/babel#8642]
## Bug Report

**Current Behavior**

Code transpiled by Babel produces no output, whereas the original code produces the expected output.

**Input Code**

Link to repo: https://github.com/ealmansi/babel-bug-sample

How to run:
- Clone repo
- Run `yarn && yarn build`
- Compare the output of the following commands:
i) `node src/index.js`
ii) `node lib/index.js`

**Expected behavior/code**

I expect both commands to produce the same output: `res [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 ]`.
The original code produces the right output, whereas the transpiled code produces no output.

**Babel Configuration (.babelrc, package.json, cli command)**

.babelrc found in repo.

**Environment**

- Babel version(s): 7.0.0
- Node/npm version: node v8.11.3, npm 5.6.0
- OS: Ubuntu 18.04.1
- Monorepo: lerna
- How you are using Babel: `cli`

**Additional context/Screenshots**

The sample code uses the async library: https://www.npmjs.com/package/async, as well as the promisify function in node's util package. Either of these could be conflicting with the runtime regenerator.

The code sample given in the repo is a minimal example. I came across this bug in a larger app where the transpiled code not only fails to produce the expected output, but also causes the app to freeze.

**Comment by loganfsmyth** [comment]
This is unfortunately because of the way the `async` module attempts to detect async functions, rather than exposing different APIs for promise-returning functions. For instance, your code also fails if you change:

```
const res = await mapLimit(numbers, 3, async function(number) {
  return number
});
```
into
```
const res = await mapLimit(numbers, 3, number => Promise.resolve(number))
```
even though they are functionally identical as functions that return promises for numbers.

We could potentially address this in Babel, so I'm alright leaving this open, but from my point of view the current `async` implementation for async functions is a design mistake.

**Comment by ealmansi** [comment]
Sorry, I did not mean to undo your changes to the title. Reverting

**Comment by ealmansi** [comment]
@loganfsmyth So, the problem here is that Babel converts an `AsyncFunction` into a non-async function that returns a promise, whereas the `async` module requires it to be an `AsyncFunction`?

**Comment by loganfsmyth** [comment]
Yeah, it does this check here: https://github.com/caolan/async/blob/db49b8922b541d011d1e457b658ef8435539fdb4/lib/internal/wrapAsync.js#L4, and Babel doesn't currently emulate the native async function `Symbol.toStringTag` property.

Realistically we can do this, so I think this is a fine bug report, but it always makes me sad to see APIs that rely on `toStringTag` since it means you can't refactor seamlessly between async functions and promises. I think functions should be treated as opaque callable objects and nothing more, otherwise you're increasing the surface area of your API a ton and asking for bugs like this.

**Comment by ealmansi** [comment]
A workaround for any visitors from the future stumbling upon this issue.

Instead of calling `eachLimit` with an async function `asyncFn`, you can pass as an argument the following: `callbackify((...args) => asyncFn(...args))`.

[] https://www.npmjs.com/package/callbackify

**Comment by danieltroger** [comment]
+1 for this bug report! I want `function.constructor.name === "AsyncFunction"` to work.

**Comment by bertho-zero** [comment]
The following code works in the Chrome console but not with a babelified code, the result of `instanceof` is always true:

```js
const AsyncFunction = Object.getPrototypeOf(async () => {}).constructor;

async function asyncMethod() {}
function method() {}

asyncMethod instanceof AsyncFunction // true
method instanceof AsyncFunction // false, but true with babelified code
```

**Comment by upupzealot** [comment]
any solution so far?
